File: external/boost_1.55.0/libs/python/test/virtual_functions.cpp
// Copyright David Abrahams 2002.
// Distributed under the Boost Software License, Version 1.0. (See
// accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)

#include <boost/python/class.hpp>
#include <boost/python/module.hpp>
#include <boost/python/def.hpp>
#include <boost/python/return_internal_reference.hpp>
#include <boost/python/call_method.hpp>
#include <boost/ref.hpp>
#include <boost/utility.hpp>

#define BOOST_ENABLE_ASSERT_HANDLER
#include <boost/assert.hpp>

using namespace boost::python;

struct X
{
    explicit X(int x) : x(x), magic(7654321) { ++counter; }
    X(X const& rhs) : x(rhs.x), magic(7654321) { ++counter; }
    virtual ~X() { BOOST_ASSERT(magic == 7654321); magic = 6666666; x = 9999; --counter; }

    void set(int _x) { BOOST_ASSERT(magic == 7654321); this->x = _x; }
    int value() const { BOOST_ASSERT(magic == 7654321); return x; }
    static int count() { return counter; }
 private:
    void operator=(X const&);
 private:
    int x;
    long magic;
    static int counter;
};

struct Y : X
{
    Y(int x) : X(x) {};
};

struct abstract : X
{
    abstract(int x) : X(x) {};
    int call_f(Y const& y) { return f(y); }
    virtual int f(Y const& y) = 0;
    abstract& call_g(Y const& y) { return g(y); }
    virtual abstract& g(Y const& y) = 0;
};

struct concrete : X
{
    concrete(int x) : X(x) {};
    int call_f(Y const& y) { return f(y); }
    virtual int f(Y const& y) { set(y.value()); return y.value(); }
};

struct abstract_callback : abstract
{
    abstract_callback(PyObject* p, int x)
        : abstract(x), self(p)
    {}

    int f(Y const& y)
    {
        return call_method<int>(self, "f", boost::ref(y));
    }

    abstract& g(Y const& y)
    {
        return call_method<abstract&>(self, "g", boost::ref(y));
    }

    PyObject* self;
};

struct concrete_callback : concrete
{
    concrete_callback(PyObject* p, int x)
        : concrete(x), self(p)
    {}

    concrete_callback(PyObject* p, concrete const& x)
        : concrete(x), self(p)
    {}

    int f(Y const& y)
    {
        return call_method<int>(self, "f", boost::ref(y));
    }

    int f_impl(Y const& y)
    {
        return this->concrete::f(y);
    }
    
    PyObject* self;
};

int X::counter;

BOOST_PYTHON_MODULE(virtual_functions_ext)
{
    class_<concrete, concrete_callback>("concrete", init<int>())
        .def("value", &concrete::value)
        .def("set", &concrete::set)
        .def("call_f", &concrete::call_f)
        .def("f", &concrete_callback::f_impl)
        ;
        
    class_<abstract, boost::noncopyable, abstract_callback
        >("abstract", init<int>())
            
        .def("value", &abstract::value)
        .def("call_f", &abstract::call_f)
        .def("call_g", &abstract::call_g, return_internal_reference<>())
        .def("set", &abstract::set)
        ;
        
    class_<Y>("Y", init<int>())
        .def("value", &Y::value)
        .def("set", &Y::set)
        ;
}

#include "module_tail.cpp"

File: external/boost_1.55.0/libs/python/test/virtual_functions.py
# Copyright David Abrahams 2004. Distributed under the Boost
# Software License, Version 1.0. (See accompanying
# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
'''
>>> from virtual_functions_ext import *

>>> class C1(concrete):
...     def f(self, y):
...         return concrete.f(self, Y(-y.value()))

>>> class C2(concrete):
...     pass

>>> class A1(abstract):
...     def f(self, y):
...         return y.value() * 2
...     def g(self, y):
...         return self

>>> class A2(abstract):
...     pass


>>> y1 = Y(16)
>>> y2 = Y(17)



#
# Test abstract with f,g overridden
#
>>> a1 = A1(42)
>>> a1.value()
42

# Call f,g indirectly from C++
>>> a1.call_f(y1)
32
>>> assert type(a1.call_g(y1)) is abstract

# Call f directly from Python
>>> a1.f(y2)
34

#
# Test abstract with f not overridden
#
>>> a2 = A2(42)
>>> a2.value()
42

# Call f indirectly from C++
>>> try: a2.call_f(y1)
... except AttributeError: pass
... else: print 'no exception'

# Call f directly from Python
>>> try: a2.call_f(y2)
... except AttributeError: pass
... else: print 'no exception'

############# Concrete Tests ############

#
# Test concrete with f overridden
#
>>> c1 = C1(42)
>>> c1.value()
42

# Call f indirectly from C++
>>> c1.call_f(y1)
-16

# Call f directly from Python
>>> c1.f(y2)
-17

#
# Test concrete with f not overridden
#
>>> c2 = C2(42)
>>> c2.value()
42

# Call f indirectly from C++
>>> c2.call_f(y1)
16

# Call f directly from Python
>>> c2.f(y2)
17


'''

def run(args = None):
    import sys
    import doctest

    if args is not None:
        sys.argv = args
    return doctest.testmod(sys.modules.get(__name__))
    
if __name__ == '__main__':
    print "running..."
    import sys
    status = run()[0]
    if (status == 0): print "Done."
    sys.exit(status)
